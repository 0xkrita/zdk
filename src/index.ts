import { GraphQLClient } from 'graphql-request';
import {
  getSdk,
  Chain,
  Network,
  NetworkInput,
  SortDirection,
  TokenInput,
  TokensQueryInput,
  TokensQueryFilter,
  CollectionsQueryInput,
  CollectionSortKeySortInput,
  TokenSortInput,
  CollectionSortKey,
  TokenSortKey,
  AggregateAttributesQueryVariables,
  OwnersByCountQueryVariables,
  SalesVolumeQueryVariables,
  EventsQueryInput,
  EventsQueryFilter,
  EventSortKeySortInput,
  EventSortKey,
  MarketSortKeySortInput,
  MarketsQueryFilter,
  MarketsQueryInput,
  MarketSortKey,
  MintsQueryFilter,
  MintsQueryInput,
  MintSortKey,
  MintSortKeySortInput,
  FloorPriceQueryVariables,
  NftCountQueryVariables,
  FullTextSearchQueryVariables,
  TokenQueryVariables,
  SaleSortKey,
  SalesQueryInput,
  SaleSortKeySortInput,
  SalesQueryFilter,
} from './queries/queries-sdk';

// Export chain and network for API users
export { Chain as ZDKChain, Network as ZDKNetwork };

export type OverrideNetworksOption = NetworkInput[];

export type OverridePaginationOptions = {
  limit?: number;
  offset?: number;
};

type SharedQueryParams = {
  networks?: NetworkInput[];
  pagination?: OverridePaginationOptions;
};

type CollectionsQueryArgs = {
  where: CollectionsQueryInput;
  includeFullDetails?: boolean;
  sort: CollectionSortKeySortInput;
} & SharedQueryParams;

type TokensQueryArgs = {
  where: TokensQueryInput;
  filter?: TokensQueryFilter;
  sort?: TokenSortInput;
  includeFullDetails?: boolean;
  includeSalesHistory?: boolean;
} & SharedQueryParams;

type EventsQueryArgs = {
  where: EventsQueryInput;
  sort?: EventSortKeySortInput;
  filter?: EventsQueryFilter;
} & SharedQueryParams;

type MarketQueryArgs = {
  where: MarketsQueryInput;
  sort: MarketSortKeySortInput;
  filter: MarketsQueryFilter;
  includeFullDetails: boolean;
} & SharedQueryParams;

type MintsQueryArgs = {
  where: MintsQueryInput;
  sort: MintSortKeySortInput;
  filter: MintsQueryFilter;
  includeFullDetails: boolean;
} & SharedQueryParams;

type SalesQueryArgs = {
  where: SalesQueryInput;
  sort: SaleSortKeySortInput;
  filter: SalesQueryFilter;
  includeFullDetails: boolean;
} & SharedQueryParams;

export interface ListOptions<SortInput> {
  networks?: OverrideNetworksOption;
  pagination?: OverridePaginationOptions;
  sort?: SortInput;
  includeFullDetails?: boolean;
}

export interface AggregateOptions {
  networks?: OverrideNetworksOption;
}

export type TokenQueryList = TokenInput;

type OptionalNetwork<K> = Omit<K, 'networks'> & {
  networks?: OverrideNetworksOption;
};

const DEFAULT_PROD_ENDPOINT = 'https://api.zora.co/graphql';

export class ZDK {
  endpoint: string;
  defaultNetworks: OverrideNetworksOption;
  defaultMaxPageSize: number = 200;

  public sdk: ReturnType<typeof getSdk>;

  constructor(
    endpoint: string = DEFAULT_PROD_ENDPOINT,
    networks: OverrideNetworksOption = [
      { network: Network.Ethereum, chain: Chain.Mainnet },
    ]
  ) {
    this.endpoint = endpoint;
    this.defaultNetworks = networks;
    this.sdk = getSdk(new GraphQLClient(this.endpoint));
  }

  private getNetworksOption = (networks?: OverrideNetworksOption) => {
    return {
      networks: networks ?? this.defaultNetworks,
    };
  };

  private getPaginationOptions = ({ limit, offset }: OverridePaginationOptions = {}) => {
    return {
      pagination: {
        limit: limit || this.defaultMaxPageSize,
        offset: offset || 0,
      },
    };
  };

  public tokens = async ({
    where,
    filter,
    pagination,
    networks,
    sort,
    includeFullDetails = false,
    includeSalesHistory = false,
  }: TokensQueryArgs) =>
    this.sdk.tokens({
      where,
      filter,
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Asc,
        sortKey: sort?.sortKey || TokenSortKey.Transferred,
        sortAxis: sort?.sortAxis,
      },
      includeFullDetails,
      includeSalesHistory,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
    });

  public token = async ({
    token,
    network = this.defaultNetworks[0],
    includeFullDetails,
  }: OptionalNetwork<TokenQueryVariables>) =>
    await this.sdk.token({
      token,
      network,
      includeFullDetails,
    });

  public events = async ({
    networks,
    filter,
    pagination,
    sort,
    where,
  }: EventsQueryArgs) =>
    this.sdk.events({
      filter,
      where,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || EventSortKey.Created,
      },
    });

  public markets = async ({
    networks,
    filter,
    pagination,
    sort,
    where,
    includeFullDetails = false,
  }: MarketQueryArgs) =>
    this.sdk.markets({
      filter,
      where,
      includeFullDetails,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || MarketSortKey.None,
      },
    });

  public mints = async ({
    networks,
    filter,
    pagination,
    sort,
    where,
    includeFullDetails = false,
  }: MintsQueryArgs) =>
    this.sdk.mints({
      filter,
      where,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || MintSortKey.None,
      },
      includeFullDetails,
    });

  public sales = async ({
    where,
    pagination,
    networks,
    sort,
    filter,
    includeFullDetails = false,
  }: SalesQueryArgs) =>
    this.sdk.sales({
      where,
      filter,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || SaleSortKey.Time,
      },
      includeFullDetails,
    });

  public collections = async ({
    where,
    pagination,
    networks,
    sort,
    includeFullDetails = false,
  }: CollectionsQueryArgs) =>
    this.sdk.collections({
      where,
      includeFullDetails,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworksOption(networks),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || CollectionSortKey.Created,
      },
    });

  public ownersByCount = async ({
    where,
    pagination,
    networks = this.defaultNetworks,
  }: OptionalNetwork<OwnersByCountQueryVariables>) =>
    this.sdk.ownersByCount({
      where,
      networks,
      pagination,
    });

  public aggregateAttributes = async ({
    where,
    networks = this.defaultNetworks,
  }: OptionalNetwork<AggregateAttributesQueryVariables>) =>
    this.sdk.aggregateAttributes({ where, networks });

  public salesVolume = async ({
    where,
    networks = this.defaultNetworks,
    timeFilter,
  }: OptionalNetwork<SalesVolumeQueryVariables>) =>
    this.sdk.salesVolume({
      where,
      networks,
      timeFilter,
    });

  public ownerCount = async ({
    where,
    networks = this.defaultNetworks,
  }: OptionalNetwork<OwnersByCountQueryVariables>) =>
    this.sdk.ownerCount({
      where,
      networks,
    });

  public floorPrice = async ({
    where,
    networks = this.defaultNetworks,
  }: OptionalNetwork<FloorPriceQueryVariables>) =>
    this.sdk.floorPrice({
      where,
      networks,
    });

  public nftCount = async ({
    where,
    networks = this.defaultNetworks,
  }: OptionalNetwork<NftCountQueryVariables>) =>
    this.sdk.nftCount({
      where,
      networks,
    });

  public search = async ({ pagination, query, filter }: FullTextSearchQueryVariables) =>
    this.sdk.fullTextSearch({
      pagination,
      filter,
      query,
    });
}
