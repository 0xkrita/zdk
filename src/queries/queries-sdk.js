'use strict';
var __makeTemplateObject =
  (this && this.__makeTemplateObject) ||
  function (cooked, raw) {
    if (Object.defineProperty) {
      Object.defineProperty(cooked, 'raw', { value: raw });
    } else {
      cooked.raw = raw;
    }
    return cooked;
  };
var __assign =
  (this && this.__assign) ||
  function () {
    __assign =
      Object.assign ||
      function (t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];
          for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
      };
    return __assign.apply(this, arguments);
  };
exports.__esModule = true;
exports.getSdk =
  exports.GetTokenDocument =
  exports.GetCollectionsDocument =
  exports.GetTokenContractDocument =
  exports.Network =
    void 0;
var graphql_tag_1 = require('graphql-tag');
var Network;
(function (Network) {
  Network['Ethereum'] = 'ETHEREUM';
  Network['Flow'] = 'FLOW';
  Network['Solana'] = 'SOLANA';
})((Network = exports.Network || (exports.Network = {})));
exports.GetTokenContractDocument = (0, graphql_tag_1['default'])(
  templateObject_1 ||
    (templateObject_1 = __makeTemplateObject(
      [
        '\n    query getTokenContract($addresses: [String!]) {\n  tokenContracts(addresses: $addresses) {\n    name\n    description\n  }\n}\n    ',
      ],
      [
        '\n    query getTokenContract($addresses: [String!]) {\n  tokenContracts(addresses: $addresses) {\n    name\n    description\n  }\n}\n    ',
      ]
    ))
);
exports.GetCollectionsDocument = (0, graphql_tag_1['default'])(
  templateObject_2 ||
    (templateObject_2 = __makeTemplateObject(
      [
        '\n    query getCollections($addresses: [String!]) {\n  collections(addresses: $addresses) {\n    name\n    description\n  }\n}\n    ',
      ],
      [
        '\n    query getCollections($addresses: [String!]) {\n  collections(addresses: $addresses) {\n    name\n    description\n  }\n}\n    ',
      ]
    ))
);
exports.GetTokenDocument = (0, graphql_tag_1['default'])(
  templateObject_3 ||
    (templateObject_3 = __makeTemplateObject(
      [
        '\n    query getToken($addresses: [String!]) {\n  tokens(addresses: $addresses) {\n    name\n    metadata\n  }\n}\n    ',
      ],
      [
        '\n    query getToken($addresses: [String!]) {\n  tokens(addresses: $addresses) {\n    name\n    metadata\n  }\n}\n    ',
      ]
    ))
);
var defaultWrapper = function (action, _operationName) {
  return action();
};
function getSdk(client, withWrapper) {
  if (withWrapper === void 0) {
    withWrapper = defaultWrapper;
  }
  return {
    getTokenContract: function (variables, requestHeaders) {
      return withWrapper(function (wrappedRequestHeaders) {
        return client.request(
          exports.GetTokenContractDocument,
          variables,
          __assign(__assign({}, requestHeaders), wrappedRequestHeaders)
        );
      }, 'getTokenContract');
    },
    getCollections: function (variables, requestHeaders) {
      return withWrapper(function (wrappedRequestHeaders) {
        return client.request(
          exports.GetCollectionsDocument,
          variables,
          __assign(__assign({}, requestHeaders), wrappedRequestHeaders)
        );
      }, 'getCollections');
    },
    getToken: function (variables, requestHeaders) {
      return withWrapper(function (wrappedRequestHeaders) {
        return client.request(
          exports.GetTokenDocument,
          variables,
          __assign(__assign({}, requestHeaders), wrappedRequestHeaders)
        );
      }, 'getToken');
    },
  };
}
exports.getSdk = getSdk;
var templateObject_1, templateObject_2, templateObject_3;
