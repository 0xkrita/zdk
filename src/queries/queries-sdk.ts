import { GraphQLClient } from 'graphql-request';
import * as Dom from 'graphql-request/dist/types.dom';
import gql from 'graphql-tag';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
  [SubKey in K]: Maybe<T[SubKey]>;
};
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  /** The GenericScalar scalar type represents a generic GraphQL scalar value that could be: List or Object. */
  JSONScalar: any;
};

export type Collection = {
  __typename?: 'Collection';
  address: Scalars['String'];
  name: Scalars['String'];
  description: Scalars['String'];
  totalSupply?: Maybe<Scalars['Int']>;
};

export enum Network {
  Ethereum = 'ETHEREUM',
  Flow = 'FLOW',
  Solana = 'SOLANA',
}

export type RootQuery = {
  __typename?: 'RootQuery';
  /** Get all collections */
  collections: Array<Collection>;
  /** Get all token contracts */
  tokenContracts: Array<TokenContract>;
  /** Get all tokens */
  tokens: Array<Token>;
};

export type RootQueryCollectionsArgs = {
  addresses?: Maybe<Array<Scalars['String']>>;
};

export type RootQueryTokenContractsArgs = {
  addresses?: Maybe<Array<Scalars['String']>>;
};

export type RootQueryTokensArgs = {
  addresses?: Maybe<Array<Scalars['String']>>;
};

export type Token = {
  __typename?: 'Token';
  tokenAddress: Scalars['String'];
  tokenId: Scalars['String'];
  network: Network;
  tokenUrl: Scalars['String'];
  tokenUrlMimeType?: Maybe<Scalars['String']>;
  owner: Scalars['String'];
  tokenContract: TokenContract;
  mintBlockNumber: Scalars['Int'];
  name?: Maybe<Scalars['String']>;
  description?: Maybe<Scalars['String']>;
  externalURL?: Maybe<Scalars['String']>;
  metadata?: Maybe<Scalars['JSONScalar']>;
  attributes?: Maybe<Array<TokenAttribute>>;
  minter?: Maybe<Scalars['String']>;
  lastRefreshTime?: Maybe<Scalars['Int']>;
};

export type TokenAttribute = {
  __typename?: 'TokenAttribute';
  traitType: Scalars['String'];
  name?: Maybe<Scalars['String']>;
  value?: Maybe<Scalars['String']>;
  role?: Maybe<Scalars['String']>;
  displayType?: Maybe<Scalars['String']>;
  maxValue?: Maybe<Scalars['String']>;
  traitCount?: Maybe<Scalars['Int']>;
  shares?: Maybe<Scalars['Int']>;
};

export type TokenContract = {
  __typename?: 'TokenContract';
  tokenAddress: Scalars['String'];
  network: Scalars['String'];
  chain: Scalars['Int'];
  name: Scalars['String'];
  symbol: Scalars['String'];
  totalSupply?: Maybe<Scalars['Int']>;
  schema: Scalars['String'];
  description: Scalars['String'];
  creator: Scalars['String'];
  iconURL: Scalars['String'];
  externalURL: Scalars['String'];
};

export const GetTokenContractDocument = gql`
  query getTokenContract($addresses: [String!]) {
    tokenContracts(addresses: $addresses) {
      name
      description
    }
  }
`;
export const GetCollectionsDocument = gql`
  query getCollections($addresses: [String!]) {
    collections(addresses: $addresses) {
      name
      description
    }
  }
`;
export const GetTokenDocument = gql`
  query getToken($addresses: [String!]) {
    tokens(addresses: $addresses) {
      name
      metadata
    }
  }
`;

export type SdkFunctionWrapper = <T>(
  action: (requestHeaders?: Record<string, string>) => Promise<T>,
  operationName: string
) => Promise<T>;

const defaultWrapper: SdkFunctionWrapper = (action, _operationName) => action();

export function getSdk(
  client: GraphQLClient,
  withWrapper: SdkFunctionWrapper = defaultWrapper
) {
  return {
    getTokenContract(
      variables?: GetTokenContractQueryVariables,
      requestHeaders?: Dom.RequestInit['headers']
    ): Promise<GetTokenContractQuery> {
      return withWrapper(
        (wrappedRequestHeaders) =>
          client.request<GetTokenContractQuery>(GetTokenContractDocument, variables, {
            ...requestHeaders,
            ...wrappedRequestHeaders,
          }),
        'getTokenContract'
      );
    },
    getCollections(
      variables?: GetCollectionsQueryVariables,
      requestHeaders?: Dom.RequestInit['headers']
    ): Promise<GetCollectionsQuery> {
      return withWrapper(
        (wrappedRequestHeaders) =>
          client.request<GetCollectionsQuery>(GetCollectionsDocument, variables, {
            ...requestHeaders,
            ...wrappedRequestHeaders,
          }),
        'getCollections'
      );
    },
    getToken(
      variables?: GetTokenQueryVariables,
      requestHeaders?: Dom.RequestInit['headers']
    ): Promise<GetTokenQuery> {
      return withWrapper(
        (wrappedRequestHeaders) =>
          client.request<GetTokenQuery>(GetTokenDocument, variables, {
            ...requestHeaders,
            ...wrappedRequestHeaders,
          }),
        'getToken'
      );
    },
  };
}
export type Sdk = ReturnType<typeof getSdk>;
export type GetTokenContractQueryVariables = Exact<{
  addresses?: Maybe<Array<Scalars['String']> | Scalars['String']>;
}>;

export type GetTokenContractQuery = { __typename?: 'RootQuery' } & {
  tokenContracts: Array<
    { __typename?: 'TokenContract' } & Pick<TokenContract, 'name' | 'description'>
  >;
};

export type GetCollectionsQueryVariables = Exact<{
  addresses?: Maybe<Array<Scalars['String']> | Scalars['String']>;
}>;

export type GetCollectionsQuery = { __typename?: 'RootQuery' } & {
  collections: Array<
    { __typename?: 'Collection' } & Pick<Collection, 'name' | 'description'>
  >;
};

export type GetTokenQueryVariables = Exact<{
  addresses?: Maybe<Array<Scalars['String']> | Scalars['String']>;
}>;

export type GetTokenQuery = { __typename?: 'RootQuery' } & {
  tokens: Array<{ __typename?: 'Token' } & Pick<Token, 'name' | 'metadata'>>;
};
