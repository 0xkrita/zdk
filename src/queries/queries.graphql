fragment CollectionFragment on Collection {
  name
  symbol
  address
  totalSupply
}

fragment FullMedia on TokenContentMedia {
size
    url
    size
    mimeType
    mediaType
    mediaEncoding {
      preview
      original
    }
}

fragment TokenSummaryFragment on Token {
  minter
  tokenId
  mintInfo {
    blockTimestamp
    blockNumber
    transactionHash
  }
  tokenUrl
  tokenAddress
  lastRefreshTime
  owner
  name
  description
  image {
    ...FullMedia
  }
  content {
    ...FullMedia
  }
}

fragment TokenFullFragment on Token {
  ...TokenSummaryFragment

  metadata

  # todo(iain): include transfer / market information?
}

query collections($network: NetworkInput!, $addresses: [String!]!, $pagination: PaginationInput!, $sort: CollectionSortKeySortInput!) {
  collections(addresses: $addresses, network: $network, pagination: $pagination, sort: $sort) {
    totalCount
    pageInfo {
      limit
      offset
    }
    nodes {
      ...CollectionFragment
    }
  }
}

query tokens(
  $network: NetworkInput!
  $addresses: [String!]!
  $pagination: PaginationInput!
  $sort: TokenSortKeySortInput!
) {
  tokens(addresses: $addresses, network: $network, pagination: $pagination, sort: $sort) {
    totalCount
    pageInfo {
      limit
      offset
    }
    nodes {
      ...TokenFullFragment
    }
  }
}

query token($network: NetworkInput!, $token: TokenInput!) {
  token(token: $token, network: $network) {
    ...TokenFullFragment
  }
}

query tokensSummary(
  $network: NetworkInput!
  $addresses: [String!]!
  $pagination: PaginationInput!
  $sort: TokenSortKeySortInput!
) {
  tokens(addresses: $addresses, network: $network, pagination: $pagination, sort: $sort) {
    totalCount
    pageInfo {
      limit
      offset
    }
    nodes {
      ...TokenSummaryFragment
    }
  }
}
