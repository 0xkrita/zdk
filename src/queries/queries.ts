import { gql } from 'graphql-request';


export const QUERY_TOKEN_CONTRACT = gql`
  query TokenContract($addresses: [String!]) {
    tokenContract(addresses: $addresses) {
      name
      description
    }
  }
`;

export const QUERY_COLLECTIONS = gql`
  query Collections($addresses: [String!]) {
    collections(addresses: $addresses) {
      name
      description
    }
  }
`;

export const QUERY_TOKEN = gql`
  query Token($addresses: [String!]) {
    token(addresses: $addresses) {
      name
      metadata
    }
  }
`;
