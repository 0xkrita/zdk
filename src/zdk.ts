// import axios from 'axios';
import { GraphQLClient } from 'graphql-request';
import {
  Chain,
  getSdk,
  Network,
  SortDirection,
  TokenSortKey,
  TokenSortKeySortInput,
} from './queries/queries-sdk';

// Export chain and network for API users
export { Chain as ZDKChain, Network as ZDKNetwork };

type OverrideNetworkOptions = {
  network?: Network;
  chain?: Chain;
};

type OverridePaginationOptions = {
  limit?: number;
  offset?: number;
};

interface ListOptions<SortInput> {
  network?: OverrideNetworkOptions;
  pagination?: OverridePaginationOptions;
  sort?: SortInput;
}

export class ZDK {
  endpoint: string;
  defaultNetwork: Network;
  defaultChain: Chain;
  defaultMaxPageSize: number = 200;

  sdk: ReturnType<typeof getSdk>;

  constructor(endpoint: string, network: Network, chain: Chain) {
    this.endpoint = endpoint;
    this.defaultNetwork = network;
    this.defaultChain = chain;
    this.sdk = getSdk(new GraphQLClient(this.endpoint));
  }

  // private fetch(url: string, options: any) {
    // return axios({url, ...options});
  // }

  getNetworkOptions = ({ network, chain }: OverrideNetworkOptions = {}) => {
    return {
      network: {
        network: network ?? this.defaultNetwork,
        chain: chain ?? this.defaultChain,
      },
    };
  };

  getPaginationOptions = ({ limit, offset }: OverridePaginationOptions = {}) => {
    return {
      pagination: {
        limit: limit || this.defaultMaxPageSize,
        offset: offset || 0,
      },
    };
  };

  token = async (
    contract: string,
    tokenId: string,
    networkOptions?: OverrideNetworkOptions
  ) =>
    this.sdk.token({
      token: { tokenId, address: contract },
      ...this.getNetworkOptions(networkOptions),
    });

  tokens = async (
    addresses: string[],
    { pagination, network, sort }: ListOptions<TokenSortKeySortInput> = {}
  ) =>
    this.sdk.tokens({
      addresses,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworkOptions(network),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || TokenSortKey.TokenId,
      },
    });

  tokensSummary = async (
    addresses: string[],
    { pagination, network, sort }: ListOptions<TokenSortKeySortInput> = {}
  ) =>
    this.sdk.tokensSummary({
      addresses,
      ...this.getPaginationOptions(pagination),
      ...this.getNetworkOptions(network),
      sort: {
        sortDirection: sort?.sortDirection || SortDirection.Desc,
        sortKey: sort?.sortKey || TokenSortKey.TokenId,
      },
    });
}
