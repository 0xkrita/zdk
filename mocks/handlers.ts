import { graphql, ResponseResolver } from 'msw';
import { getSdk } from '../src/queries/queries-sdk';
import { makeServer } from './response';

const handleGenericQuery: ResponseResolver<any> = async (req, res, ctx) => {
  const server = makeServer({}, {});
  const serverResp = await server.query(req.body?.query, req.body?.variables);
  return res((ctx as any).data(serverResp));
};

// @ts-ignore
const queries = Object.keys(getSdk(null, null));

export const handlers = queries.map((queryName) =>
  graphql.query(queryName, handleGenericQuery)
);
