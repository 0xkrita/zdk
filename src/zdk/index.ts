import axios from 'axios';
import { GraphQLClient } from 'graphql-request';
import { Chain, getSdk, Network, SortDirection } from '../queries/queries-sdk';

type OverrideNetworkOptions = {
  network?: Network;
  chain?: Chain;
};

type OverridePaginationOptions = {
  limit: number;
  offset: number;
  sortKey: string;
  sortDirection: SortDirection;
};

export class ZDK {
  endpoint: string;
  defaultNetwork: Network;
  defaultChain: Chain;
  defaultMaxPageSize: number = 200;

  sdk: ReturnType<typeof getSdk>;

  constructor(endpoint: string, network: Network, chain: Chain) {
    this.endpoint = endpoint;
    this.defaultNetwork = network;
    this.defaultChain = chain;
    this.sdk = getSdk(new GraphQLClient(this.endpoint, { fetch: axios }));
  }

  getNetworkOptions = ({ network, chain }: OverrideNetworkOptions = {}) => {
    return {
      network: {
        network: network ?? this.defaultNetwork,
        chain: chain ?? this.defaultChain,
      },
    };
  };

  getPaginationOptions = (
    { limit, offset, sortKey, sortDirection }: OverridePaginationOptions = {limit: this.defaultMaxPageSize, offset: 0, sortKey: '', sortDirection: SortDirection.Desc}, defaultSortKey: string
  ) => {
    return {
      pagination: {
        limit,
        offset,
        sortKey: sortKey || defaultSortKey,
        sortDirection: sortDirection || SortDirection.Desc,
      }
    };
  };

  token = async (
    contract: string,
    tokenId: string,
    networkOptions?: OverrideNetworkOptions
  ) =>
    this.sdk.token({
      token: { tokenId, address: contract },
      ...this.getNetworkOptions(networkOptions),
    });

  tokenSummary = async (
    contract: string,
    tokenId: string,
    networkOptions?: OverrideNetworkOptions
  ) =>
    this.sdk.tokenSummary({
      token: { tokenId, address: contract },
      ...this.getNetworkOptions(networkOptions),
    });

  tokens = async (
    addresses: string[],
    paginationOptions?: OverridePaginationOptions,
    networkOptions?: OverrideNetworkOptions
  ) =>
    this.sdk.tokens({
      addresses,
      ...this.getPaginationOptions(paginationOptions, 'tokenId'),
      ...this.getNetworkOptions(networkOptions),
    });

  tokensSummary = async (
    addresses: string[],
    paginationOptions?: OverridePaginationOptions,
    networkOptions?: OverrideNetworkOptions
  ) =>
    this.sdk.tokensSummary({
      addresses,
      ...this.getPaginationOptions(paginationOptions, 'tokenId'),
      ...this.getNetworkOptions(networkOptions),
    });
}
