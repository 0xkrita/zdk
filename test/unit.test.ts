import { registry } from '../mocks/handlers';
import { Chain, Network } from '../src/queries/queries-sdk';
import { ZDK } from '../src/zdk';
import './setup-mocks';

const ZORA_TESTING_PATH = 'http://example.com/graphql';

describe('unit zdk', () => {
  let zdk: ZDK;
  let querySpy: any;
  beforeEach(() => {
    querySpy = spyOn(registry, 'query');
    zdk = new ZDK(ZORA_TESTING_PATH, Network.Ethereum, Chain.Mainnet);
  });

  it('should fetch mock token', async () => {
    const token = await zdk.token('0xCa21d4228cDCc68D4e23807E5e370C07577Dd152', '12');
    expect(token).toMatchSnapshot();
    expect(querySpy).toBeCalledWith('token', {
      network: {
        network: 'ETHEREUM',
        chain: 'MAINNET',
      },
      token: { tokenId: '12', address: '0xCa21d4228cDCc68D4e23807E5e370C07577Dd152' },
    });
  });
  it('should fetch mock tokens', async () => {
    const tokens = await zdk.tokens(['0xCa21d4228cDCc68D4e23807E5e370C07577Dd152'], {
      pagination: { limit: 2 },
    });
    expect(tokens.tokens.nodes.length).toBe(2);
    expect(tokens.tokens.totalCount).toBe(2);
    expect(querySpy).toBeCalledWith('tokens', {
      pagination: { limit: 2, offset: 0 },
      sort: { sortDirection: 'DESC', sortKey: 'TOKEN_ID' },
      network: {
        network: 'ETHEREUM',
        chain: 'MAINNET',
      },
      addresses: ['0xCa21d4228cDCc68D4e23807E5e370C07577Dd152'],
    });
  });
});
