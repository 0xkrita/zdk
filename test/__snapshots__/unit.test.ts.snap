// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit zdk should fetch mock token 1`] = `
Object {
  "token": Object {
    "__typename": "TokenWithFullMarketHistory",
    "markets": Array [
      Object {
        "collectionAddress": "Hello World",
        "marketAddress": "Hello World",
        "marketType": "V2_AUCTION",
        "networkInfo": Object {
          "chain": "MAINNET",
          "network": "ETHEREUM",
        },
        "price": Object {
          "blockNumber": 7,
          "ethPrice": Object {
            "decimal": -7.4796289074247255,
            "raw": "Hello World",
          },
          "nativePrice": Object {
            "currency": Object {
              "address": "Hello World",
              "decimals": 8,
              "name": "Hello World",
            },
            "decimal": 65.18361761440886,
            "raw": "Hello World",
          },
          "usdcPrice": Object {
            "decimal": -46.67760211342089,
            "raw": "Hello World",
          },
        },
        "status": "Hello World",
        "transactionInfo": Object {
          "blockNumber": 5,
          "blockTimestamp": 1644379949,
          "logIndex": 6,
          "transactionHash": "Hello World",
        },
      },
      Object {
        "collectionAddress": "Hello World",
        "marketAddress": "Hello World",
        "marketType": "V3_ASK",
        "networkInfo": Object {
          "chain": "MAINNET",
          "network": "ETHEREUM",
        },
        "price": Object {
          "blockNumber": 11,
          "ethPrice": Object {
            "decimal": -45.47765839908591,
            "raw": "Hello World",
          },
          "nativePrice": Object {
            "currency": Object {
              "address": "Hello World",
              "decimals": 12,
              "name": "Hello World",
            },
            "decimal": -60.067615084450445,
            "raw": "Hello World",
          },
          "usdcPrice": Object {
            "decimal": -58.42496482187158,
            "raw": "Hello World",
          },
        },
        "status": "Hello World",
        "transactionInfo": Object {
          "blockNumber": 9,
          "blockTimestamp": 1644379949,
          "logIndex": 10,
          "transactionHash": "Hello World",
        },
      },
    ],
    "token": Object {
      "collectionAddress": "Hello World",
      "content": Object {
        "mediaEncoding": Object {
          "__typename": "AudioEncodingTypes",
          "large": "Hello World",
          "original": "Hello World",
        },
        "mimeType": "Hello World",
        "size": "Hello World",
        "url": "Hello World",
      },
      "description": "Hello World",
      "image": Object {
        "mediaEncoding": Object {
          "__typename": "VideoEncodingTypes",
          "large": "Hello World",
          "original": "Hello World",
          "poster": "Hello World",
          "preview": "Hello World",
          "thumbnail": "Hello World",
        },
        "mimeType": "Hello World",
        "size": "Hello World",
        "url": "Hello World",
      },
      "lastRefreshTime": null,
      "mintInfo": Object {
        "mintContext": Object {
          "__typename": "TransactionInfo",
          "blockNumber": 3,
          "blockTimestamp": 1644379949,
          "logIndex": 4,
          "transactionHash": "Hello World",
        },
        "originatorAddress": "Hello World",
        "price": Object {
          "__typename": "PriceAtTime",
          "blockNumber": 1,
          "ethPrice": Object {
            "decimal": -6.459742808931551,
            "raw": "Hello World",
          },
          "nativePrice": Object {
            "currency": Object {
              "address": "Hello World",
              "decimals": 2,
              "name": "Hello World",
            },
            "decimal": 47.47667128119534,
            "raw": "Hello World",
          },
          "usdcPrice": Object {
            "decimal": -75.04713234180045,
            "raw": "Hello World",
          },
        },
        "toAddress": "Hello World",
      },
      "name": "Hello World",
      "owner": "Hello World",
      "tokenContract": Object {
        "chain": 0,
        "collectionAddress": "Hello World",
        "description": "Hello World",
        "name": "Hello World",
        "network": "Hello World",
        "symbol": "Hello World",
      },
      "tokenId": "Hello World",
    },
  },
}
`;
